src/tests/test_numerictype_cast.cpp

```diff
@@ -0,0 +1,118 @@
+#include <memory>
+
+#include <gtest/gtest.h>
+
+#include "types.h"
+
+namespace tests {
+struct NumericCast : public ::testing::Test {
+protected:
+    void SetUp() override {
+        dbl_b_ptr.reset(new model::Double(2.79));
+        dbl_a_ptr.reset(new model::Double(5.58));
+        int_a_ptr.reset(new model::Int(21));
+        int_b_ptr.reset(new model::Int(7));
+    }
+    void TearDown() override {
+        dbl_b_ptr.reset(new model::Double(2.79));
+        dbl_a_ptr.reset(new model::Double(5.58));
+        int_a_ptr.reset(new model::Int(21));
+        int_b_ptr.reset(new model::Int(7));
+    }
+    std::unique_ptr<model::Double> dbl_a_ptr;
+    std::unique_ptr<model::Double> dbl_b_ptr;
+    std::unique_ptr<model::Int> int_a_ptr;
+    std::unique_ptr<model::Int> int_b_ptr;
+
+    model::DoubleType dbl_type;
+    model::INumericType& dbl_type_ref = dbl_type;
+
+    model::IntType int_type;
+    model::INumericType& int_type_ref = int_type;
+
+    std::byte* val;
+};
+TEST_F(NumericCast, GetValueAsDouble) {
+    val = reinterpret_cast<std::byte*>(dbl_a_ptr.get());
+    ASSERT_DOUBLE_EQ(dbl_type_ref.GetValueAs<model::Double>(val), *dbl_a_ptr.get());
+    val = reinterpret_cast<std::byte*>(int_a_ptr.get());
+    ASSERT_DOUBLE_EQ(int_type_ref.GetValueAs<model::Double>(val), *int_a_ptr.get());
+}
+TEST_F(NumericCast, GetValueAsInt) {
+    val = reinterpret_cast<std::byte*>(dbl_a_ptr.get());
+    ASSERT_EQ(dbl_type_ref.GetValueAs<model::Int>(val), (model::Int)*dbl_a_ptr.get());
```

Comment:
C-style casts [are forbidden](https://google.github.io/styleguide/cppguide.html#Casting) by the style guide, use `static_cast` instead